### DISCUSSION TITLE
TESTed: failed tests with functions

### DISCUSSION THREAD

**Discussion by marcothijs** [dodona-edu/feedback-and-support#33] [Questions - Vragen] [answered]
Hi, 

Please check https://dodona.be/nl/activities/638195380/
I just try to test some of the functions. Added "if __name__ == '__main__':" for the main script so it wouldn't run while testing the functions. I get 2 types of errors depending on the YAML lines
1) AttributeError: module 'submission' has no attribute 'ANDPoort' -> when I use:
- expression: "AND_poort(0, 0)"
   return: 0

2) NameError: name 'AND_poort' is not defined -> when I use:
- expression:
       python: "AND_poort(1, 1)"
   return: 0

What am I doing wrong this time...

Thanks for your help.

**Comment by niknetniko** [comment, marked as answer]
I think the blame lies 100% with TESTed this time.

There are two factors here:

- Currently, TESTed expects identifiers (like function names) to be in snake_case in the test suite and will always convert it when executing. That is why `AND_poort` is converted to `ANDPoort`, which doesn't exist in the submission. We want to support an override for this behaviour in the future, but are not there yet (https://github.com/dodona-edu/universal-judge/issues/326)
- When using the language-specific things (e.g. `python: "blabla"`), you need to manually put the submission module in the string (see https://github.com/dodona-edu/universal-judge/issues/423). So it should be `python: "submission.AND_poort(1, 1)"`. However, this will then also be shown on Dodona, so this is not ideal.

So, I think there are two choices at the moment:

- Rename the functions in the submission to `and_poort`.
- Use language-specific stuff, but with submission.

So either this:

```yaml
- expression: "and_poort(0, 0)"
  return: 0
```

Or this:

```yaml
- expression:
    python: "submission.AND_poort(0, 0)"
  return: 0
```

Thanks again for using TESTed, it helps us a lot by seeing other people's perspectives on using TESTed.

**Reply by marcothijs** [reply]
Thanks again... I will go for option 1 and keep this in mind for future excercises. Normally I use snake-case for function names (PEP-8), but this was an exception because of the digitale gate name. Further, I recall a remark on the submission.function_name thing for language-specific stuff but I didn't make the connection with the issue today. 
Anyhow, thanks again and as long as your team doesn't mind answering my .... questions, I'll keep on sending them ;-).
